### PR TITLE
Py2 and Py3 support

### PR DESCRIPTION
**Pull request by petebachant**
Basically this is just #260 rebased onto the current `develop`.  I suspect some changes to `TestCase.py` may have been lost while fixing merge conflicts.

### CODE REVIEW

**Comment by petebachant** [comment]
The Travis config should probably be updated to use Miniconda, running tests with both a Python 2 and Python 3 environment. Could take some influence from https://github.com/mwaskom/seaborn/blob/master/.travis.yml

**Comment by petebachant** [comment]
Fixed at least one failed test, but there are still many print statements left to fix:

image

Hopefully this can be fixed up and merged before a bunch of new Python 3 incompatible code is added!

**Comment by economon** [comment]
Thanks for spending the time on this, @petebachant. I just fixed a couple of bugs in the python scripts in another PR, but I don't think they will affect this too much.

Btw, do you have access to a sizable distributed machine with Python 3+ installed for testing? It would be great to know if this is working in a production environment, while keeping backward compatibility with Python 2+, since it can sometimes be hard to get new Python installations on clusters.

**Comment by petebachant** [comment]
I don't have access to a cluster to test this on. However, I usually will install Anaconda Python (3) in my home directory when I do run on clusters just to have complete control over the entire Python environment. I think for scientific/engineering computing this is generally a better strategy than relying on the system Python. 

On another note, any hints on why these tests are failing?

```
/home/travis/build/su2code/SU2/TestCases/optimization_euler/steady_naca0012
naca0012_geo: FAILED
```

```
execution command: SU2_GEO inv_NACA0012_adv.cfg > inv_NACA0012_adv.log
ERROR: The code was not able to get to the "OBJFUN" section.
ERROR: The SU2_GEO values could not be found.
```

```
/home/travis/build/su2code/SU2/TestCases/rans/naca0012
Error in test_vals!
turb_naca0012_sst_restart_mg: FAILED
```

**Comment by talbring** [comment]
I don't know if this is related, but the output of the stored test vals looks a little bit strange now:

```
==================== Start Test: contadj_fixedcl_naca0012 ====================
/home/travis/build/su2code/SU2/TestCases/fixed_cl/naca0012
contadj_fixedcl_naca0012: PASSED
execution command: parallel_computation.py -f inv_NACA0012_ContAdj.cfg.autotest > inv_NACA0012_ContAdj.cfg.log
test_iter=500 
 test_vals (stored):  -2.093175,
-6.967350,
0.249870,
-0.000019,
sim_vals (computed):  -2.093175, -6.967350, 0.249870, -0.000019, 
 delta_vals:  0.000000, 0.000000, 0.000000, 0.000000, 
test duration: 0.62 min
==================== End Test: contadj_fixedcl_naca0012 ====================
```

**Comment by petebachant** [comment]
@talbring it looks like it is related. I will try to fix that.

**Comment by economon** [comment]
On the topic of Python distributions: I agree.. I typically build up a local Enthought Canopy Python (command line only) on clusters to have control too. Although, it seems that their new versions (canopy 2+) deprecate the command line interface somewhat, even though Py 3+ is available. Perhaps I'll give Anaconda a try then soon, thanks for that.

File: TestCases/TestCase.py [hunk 1]
```diff
@@ -169,22 +169,25 @@ def run_test(self):
...
-
-        print('delta_vals: ', end=" ")
-        for j in delta_vals:
-            print('%f,'%j, end=" ")
-        print()
+        print('test_iter=%d' % self.test_iter)
+
+        print(
+            'test_vals (stored): ' + ', '.join(
+                '{:f}'.format(j) for j in test_vals
```
Comment (petebachant) [inline TestCases/TestCase.py]: Should be `self.test_vals`

File: TestCases/TestCase.py [hunk 2]
```diff
@@ -360,22 +363,25 @@ def run_opt(self):
...
-        print('test_iter=%d \n'%self.test_iter, end=" ")
+        print('test_iter=%d' % self.test_iter)
 
-        print('test_vals (stored): ', end=" ")
-        for j in self.test_vals:
-            print('%f,'%j, end=" ")
-        print()
+        print(
+            'test_vals (stored): ' + ', '.join(
+                '{:f}'.format(j) for j in test_vals
```
Comment (petebachant) [inline TestCases/TestCase.py]: Should be `self.test_vals`.

File: TestCases/TestCase.py [hunk 3]
```diff
@@ -504,20 +510,23 @@ def run_geo(self):
         if iter_missing:
             print('ERROR: The SU2_GEO values could not be found.')
 
-        print('test_vals (stored): ', end=" ")
-        for j in self.test_vals:
-            print('%f,'%j, end=" ")
-        print()
+        print(
+            'test_vals (stored): ' + ', '.join(
+                '{:f}'.format(j) for j in test_vals
```
Comment (petebachant) [inline TestCases/TestCase.py]: Should be `self.test_vals`.

File: TestCases/TestCase.py [hunk 4]
```diff
@@ -627,22 +636,26 @@ def run_def(self):
...
-    
-        print('delta_vals: ', end=" ")
-        for j in delta_vals:
-            print('%e'%j, end=" ")
-        print()
+        print('test_iter=%d' % self.test_iter)
+
+        print(
+            'test_vals (stored): ' + ', '.join(
+                '{:f}'.format(j) for j in test_vals
```
Comment (petebachant) [inline TestCases/TestCase.py]: Should be `self.test_vals`.

**Comment by petebachant** [comment]
Looks like there may be one issue with Python 3 compatibility: The Makefile for `pySU2` is hard-coded to used the system Python 2.7:

```makefile
PYTHON_INCLUDE = -I/usr/include/python2.7
#NUMPY_INCLUDE = /usr/lib/python2.7/dist-packages/numpy/core/include
MPI4PY_INCLUDE = ${HOME}/.local/lib/python2.7/site-packages/mpi4py/include \
                 -I/usr/local/lib/python2.7/site-packages/mpi4py/include \
                 -I/usr/local/lib/python2.7/dist-packages/mpi4py/include \
                 -I/usr/local/lib/python2.7/dist-packages/mpi4py/include \
                 -I/Library/Python/2.7/site-packages/mpi4py/include

SWIG = swig

pySU2_INCLUDE = -I${abs_top_srcdir}/Common/include \
	-I${abs_top_srcdir}/SU2_CFD/include

PY_INCLUDE = ${PYTHON_INCLUDE} -I${MPI4PY_INCLUDE}

PY_LIB = -lpython2.7 \
         -L/usr/lib \
         -L/usr/lib/x86_64-linux-gnu
```

This should probably be generalized to use whatever Python environment is currently active.

**Comment by petebachant** [comment]
So I have setup Travis to test both Python versions, but I need to learn a little bit more about automake to setup the `pySU2` Makefile properly to build against the active Python environment. Hopefully getting closer...

**Comment by petebachant** [comment]
Think I'm going to need some help fixing up the `pySU2` Makefile. I'm not familiar with how this build system works. I modified `Makefile.in` but it doesn't seem to be affecting where the compiler is looking for libs, headers, etc. for Python.

**Comment by economon** [comment]
@petebachant thanks for the effort to keep this PR going. I think that @tobadavid would be the best contact for updating the pySU2 wrapper build, as he is the original author. David, if you have a moment, could you please give this a look? Thanks!

**Comment by petebachant** [comment]
Made some progress on this. The build system now automatically finds the Python directories and library, so the wrapper is now built against both Python 2.7 and 3.6 in the test suite. However, some tests are still failing. Any ideas what might be going wrong?

**Comment by petebachant** [comment]
It looks like all the tests are finally working. However, the `directdiff_euler_py` output variables are written in the wrong order. It looks like this comes from `SU2/eval/gradients.py`:

```python
    # initialize gradients
    func_keys = list(su2io.grad_names_map.keys())
    func_keys = ['VARIABLE'] + func_keys
```

My guess is that there was some assumption that dictionary keys have an inherent order. Now, should the test be doing a diff of the text output, or should it really be checking the numerical values?

**Comment by petebachant** [comment]
Alright, so my above hunch was correct: that there was an assumption of dictionary key order. I switched that variable over to an ordered bunch and I believe all the tests are passing, despite one of them timing out.

After addressing that, the entire repo (including the wrapper) should now be compatible with both Python 2 and Python 3, and is setup to be tested with both.

**Comment by petebachant** [comment]
This PR finally passed the tests! Can I get some review feedback @economon and/or @talbring? It would be great to have this merged ASAP so no one adds any new code incompatible with Python 3.

**Comment by petebachant** [comment]
Some additional motivation: https://github.com/numpy/numpy/blob/master/doc/neps/dropping-python2.7-proposal.rst

File: externals/autotools/share/autoconf/autom4te.cfg [hunk 5]
```diff
@@ -106,7 +106,7 @@ end-language: "Autoreconf-preselections"
 # This intermediate language is used by aclocal to build aclocal.m4.
 
 begin-language: "Autoconf-without-aclocal-m4"
-args: --prepend-include '/Users/economon/SU2/externals/autotools/share/autoconf'
```
Comment (talbring) [inline externals/autotools/share/autoconf/autom4te.cfg]: Maybe you can revert those changes in the autotools directory.

File: SU2_PY/pySU2/Makefile.am [hunk 6]
```diff
@@ -42,12 +42,10 @@ SWIG_SRC = ${abs_top_srcdir}/SU2_PY/pySU2/pySU2.i
 SWIG_WRAP_REAL = SU2_APIPYTHON_wrap
 SWIG_SO_REAL = _pysu2.${SO_EXT}
 
-PYTHON_INCLUDE = -I/usr/include/python2.7
+PYTHON_SITE_PACKAGES=$(shell python -c "import site; print(site.getsitepackages()[0])")
```
Comment (tobadavid) [inline SU2_PY/pySU2/Makefile.am]: Thanks for having fixed the hard-coded search for Python and mpi4py !

File: TestCases/TestCase.py [hunk 7]
```diff
@@ -3,18 +3,16 @@
 ## \file TestCase.py
 #  \brief Python class for automated regression testing of SU2 examples
 #  \author A. Aranake, A. Campos, T. Economon, T. Lukaczyk, S. Padron
-#  \version 5.0.0 "Raven"
+#  \version 5.0 "Raven"
 #
-# SU2 Original Developers: Dr. Francisco D. Palacios.
-#                          Dr. Thomas D. Economon.
+# SU2 Lead Developers: Dr. Francisco Palacios (Francisco.D.Palacios@boeing.com).
```
Comment (economon) [inline TestCases/TestCase.py]: Looks like a few things got accidentally changed in this header.. could you please revert for consistency?

File: configure.ac [hunk 8]
```diff
@@ -56,6 +56,27 @@ AC_PROG_RANLIB
 AC_PROG_MKDIR_P
 AC_PROG_CXX
 
+# Python configuration
```
Comment (economon) [inline configure.ac]: There are indeed quite a few changes to the configure/autostuff. Anything of consequence that folks should be aware of, such as new configure options, paths, etc.?

File: externals/autotools/share/autoconf/autom4te.cfg (hunk 5 above)
Comment (petebachant) [reply externals/autotools/share/autoconf/autom4te.cfg]: Do you mean changing all those home directories back to `economon`? I think all of this happened automatically when I changed the config to look for the proper Python dirs.

Comment (talbring) [reply externals/autotools/share/autoconf/autom4te.cfg]: I am referring to all of the changes in the externals/autotools folder.

**Comment by petebachant** [comment]
Unfortunately it looks like the serial regression test with Python wrapper is timing out for both versions. Any ideas to speed this up or perhaps request another minute or two?

**Comment by talbring** [comment]
I think we can reduce the iterations of some of the test cases. That should already save a lot time ...

**Comment by petebachant** [comment]
Alright, I'll keep an eye out for that PR getting merged and will merge it into this one.

**Comment by timjim333** [comment]
Hi, I was having a few issues installing the Python wrapper (which strangely happened to work for some of my systems) and I followed the error back to SU2_PY/pySU2/Makefile.am, where a Python 2.7 path is still hardcoded. Is this an intentional remainder?

I tried to use the [fix documented back in V5](https://github.com/su2code/SU2/issues/447) but I've not had any success yet. I'm still troubleshooting (and may have a few incorrect paths still) but just to hear the official word here before I have another go. Thanks.

**Comment by petebachant** [comment]
@timjim333 you should probably open a new issue for this, specifying your system, Python environment, and the error messages you get.

It may be possible to remove those hardcoded paths, changing

```makefile
MPI4PY_INCLUDE = ${HOME}/.local/lib/python2.7/site-packages/mpi4py/include \
                 -I${PYTHON_SITE_PACKAGES}/mpi4py/include \
                 -I/Library/Python/2.7/site-packages/mpi4py/include
```

to

```makefile
MPI4PY_INCLUDE = ${PYTHON_SITE_PACKAGES}/mpi4py/include
```

**Pull request merged**